### PR TITLE
Support nodeset <switch> osimage command for cumulus switch

### PR DESCRIPTION
**Pull request by cxhong**
for issue #5126

unit test results:
````
[root@boston02 xCAT_plugin]# lsdef mid08tor03 -i provmethod
Object name: mid08tor03
    provmethod=
[root@boston02 xCAT_plugin]# nodeset mid08tor03 osimage=cumulus3.5.2-armel
mid08tor03: install cumulus3.5.2-armel
[root@boston02 xCAT_plugin]# lsdef mid08tor03 -i provmethod
Object name: mid08tor03
    provmethod=cumulus3.5.2-armel

# lsdef -t osimage cumulus3.5.2-armel
Object name: cumulus3.5.2-armel
    description=Cumulus Linux
    imagetype=linux
    osarch=armel
    osname=cumulus
    osvers=cumulus3.5.2
    pkgdir=/install/cumulus3.5.2/armel/cumulus-linux-3.5.2-bcm-armel.bin
    provmethod=install
````

The default URL updated in the dhcp leases file as pkgdir:
````
host mid08tor03 {
  dynamic;
  hardware ethernet 8c:ea:1b:e8:78:c0;
  uid 8c:ea:1b:e8:78:c0;
  fixed-address 172.21.208.3;
        supersede server.ddns-hostname = "mid08tor03";
        supersede host-name = "mid08tor03";
        if substring (option vendor-class-identifier, 0, 11) = "onie_vendor" {
          supersede www-server =
                                 "http://172.21.253.31/install/cumulus3.5.2/armel/cumulus-linux-3.5.2-bcm-armel.bin";
        }
}
````

### CODE REVIEW

**Comment by immarvin** [comment]
hi @cxhong , the PR https://github.com/xcat2/xcat-core/pull/5122 also include a `onie.pm`, which one should be merged? should the `onie.pm` be removed from this PR? thanks

File: xCAT-server/lib/xcat/plugins/dhcp.pm [hunk 1]
```diff
@@ -693,7 +693,11 @@ sub addnode
         } elsif ($nrent and $nrent->{netboot} and $nrent->{netboot} eq 'petitboot') {
             $lstatements = 'option conf-file \"http://' . $nxtsrv . '/tftpboot/petitboot/' . $node . '\";' . $lstatements;
         } elsif ($nrent and $nrent->{netboot} and $nrent->{netboot} eq 'onie') {
-            $lstatements = 'if substring (option vendor-class-identifier,0,11) = \"onie_vendor\" { option www-server = \"http://' . $nxtsrv . $ntent->{provmethod} . '\";}' . $lstatements;
+            my $provmethod = $ntent->{provmethod};
+            my $linuximagetab = xCAT::Table->new('linuximage');
+            my $imagetab = $linuximagetab->getAttribs({ imagename => $provmethod }, 'pkgdir');
+            my $image_pkgdir = $imagetab->{'pkgdir'};
```
Comment (immarvin) [inline xCAT-server/lib/xcat/plugins/dhcp.pm]: the `pkgdir` is designed to be a comma `,` delimited list of full path to file/dirs, only the full path to the bin file should be specified here. need to add logic to find the correct entry for the file full path, try `-f <path to file/dir>`

File: xCAT-server/lib/xcat/plugins/dhcp.pm [hunk 2]
```diff
@@ -693,7 +693,11 @@ sub addnode
         } elsif ($nrent and $nrent->{netboot} and $nrent->{netboot} eq 'petitboot') {
             $lstatements = 'option conf-file \"http://' . $nxtsrv . '/tftpboot/petitboot/' . $node . '\";' . $lstatements;
         } elsif ($nrent and $nrent->{netboot} and $nrent->{netboot} eq 'onie') {
-            $lstatements = 'if substring (option vendor-class-identifier,0,11) = \"onie_vendor\" { option www-server = \"http://' . $nxtsrv . $ntent->{provmethod} . '\";}' . $lstatements;
+            my $provmethod = $ntent->{provmethod};
+            my $linuximagetab = xCAT::Table->new('linuximage');
+            my $imagetab = $linuximagetab->getAttribs({ imagename => $provmethod }, 'pkgdir');
```
Comment (immarvin) [inline xCAT-server/lib/xcat/plugins/dhcp.pm]: need some logic to check whether `$provmethod` and `$image_pkgdir` is valid here , and throw messages if any.

**Comment by cxhong** [comment]
@immarvin  PR#5122 should merge first.  This PR has few changes on top.

**Comment by cxhong** [comment]
1) without provmethod
```
# chdef mid08tor03  provmethod=
# makedhcp mid08tor03
Warning: provmethod is not defined for mid08tor03
```
2) with provmethod but not in the osimage table
```
# chdef mid08tor03  provmethod=cumulus3.5.1-armel
1 object definitions have been created or modified.
# makedhcp mid08tor03
Warning: osimage cumulus3.5.1-armel is not defined in the osimage table
````
3) osimage is available but path of pkgdir is not available
```
# chdef -t osimage cumulus3.5.2-armel  pkgdir=/install/cumulus3.5.1/armel/cumulus-linux-3.5.2-bcm-armel.bin
1 object definitions have been created or modified.
[root@boston02 xCAT_plugin]# ls -ltr /install/cumulus3.5.1/armel/cumulus-linux-3.5.2-bcm-armel.bin
ls: cannot access /install/cumulus3.5.1/armel/cumulus-linux-3.5.2-bcm-armel.bin: No such file or directory
# makedhcp mid08tor03
Warning: osimage cumulus3.5.2-armel pkgdir doesn't exists
````
If there are multiple pkgdir, will pick the first available one
```
# chdef -t osimage cumulus3.5.2-armel  pkgdir=/install/cumulus3.5.1/armel/cumulus-linux-3.5.2-bcm-armel.bin,/install/cumulus3.5.2/armel/cumulus-linux-3.5.2-bcm-armel.bin
# makedhcp mid08tor03
# in the dhcpd.lease file
host mid08tor03 {
  dynamic;
  hardware ethernet 8c:ea:1b:e8:78:c0;
  uid 8c:ea:1b:e8:78:c0;
  fixed-address 172.21.208.3;
        supersede server.ddns-hostname = "mid08tor03";
        supersede host-name = "mid08tor03";
        if substring (option vendor-class-identifier, 0, 11) = "onie_vendor" {
          supersede www-server =
                                 "http://172.21.253.31/install/cumulus3.5.2/armel/cumulus-linux-3.5.2-bcm-armel.bin";
        }
}
```

File: xCAT-server/lib/xcat/plugins/dhcp.pm (hunk 2 above)
Comment (cxhong) [reply xCAT-server/lib/xcat/plugins/dhcp.pm]: added

File: xCAT-server/lib/xcat/plugins/dhcp.pm (hunk 1 above)
Comment (cxhong) [reply xCAT-server/lib/xcat/plugins/dhcp.pm]: added

**Comment by immarvin** [comment]
hi @cxhong ，please resolve the conflict

**Comment by immarvin** [comment]
agree to merge

**Pull request merged**